tools/autograd/gen_variable_type.py

```diff
@@ -654,44 +611,108 @@ def emit_record_trace(env, declaration):
         combined = nested_dict(local, nested_dict(env, declaration))
         return RECORD_TRACE.substitute(combined)
 
-    def process_function(declaration):
-        if skip_function(declaration['name']):
-            return
+    def emit_check_no_requires_grad(tensor_args, args_with_derivatives):
+        """Checks that arguments without derivatives don't require grad"""
+        body = []
+        for arg in tensor_args:
+            name = arg['name']
+            if name in args_with_derivatives:
+                continue
+            if name == 'output':
+                # Double-backwards definitions sometimes take in 'input' and
+                # 'output', but only define the derivative for input.
+                continue
+            if arg['dynamic_type'] in {'IndexTensor', 'BoolTensor'}:
+                continue
```

Comment:
You should still check that these don't require grad. We should add checks to Variable constructor (in Python), to make sure people never create index or bool tensors that do

Reply:
Our tests don't handle that correctly currently. I've added an issue to track this:
https://github.com/pytorch/pytorch/issues/3226